masonry/src/render_root.rs

```diff
@@ -295,56 +306,73 @@ impl RenderRoot {
         }
     }
 
+    /// Get a [`WidgetMut`] to the root widget.
+    ///
+    /// Because of how `WidgetMut` works, it can only be passed to a user-provided callback.
     pub fn edit_root_widget<R>(
         &mut self,
         f: impl FnOnce(WidgetMut<'_, Box<dyn Widget>>) -> R,
     ) -> R {
-        let mut fake_widget_state =
-            WidgetState::new(self.root.id(), Some(self.get_kurbo_size()), "<root>");
-        let root_state_token = self.widget_arena.widget_states.root_token_mut();
-        let root_widget_token = self.widget_arena.widgets.root_token_mut();
-        let state_ref = root_state_token
-            .into_child_mut(self.root.id().to_raw())
-            .expect("root widget not in widget tree");
-        let widget_ref = root_widget_token
-            .into_child_mut(self.root.id().to_raw())
-            .expect("root widget not in widget tree");
+        let mut root_state = WidgetState::root(self.root.id(), self.get_kurbo_size());
 
-        // Our WidgetArena stores all widgets as Box<dyn Widget>, but the "true"
-        // type of our root widget is *also* Box<dyn Widget>. We downcast so we
-        // don't add one more level of indirection to this.
-        let widget = widget_ref
-            .item
-            .as_mut_dyn_any()
-            .downcast_mut::<Box<dyn Widget>>()
-            .unwrap();
+        // TODO - Factor out into a "pre-event" function?
+        self.state.next_focused_widget = self.state.focused_widget;
+
+        let mut res = None;
+        mutate_widget(self, self.root.id(), |mut widget_mut| {
+            // Our WidgetArena stores all widgets as Box<dyn Widget>, but the "true"
+            // type of our root widget is *also* Box<dyn Widget>. We downcast so we
+            // don't add one more level of indirection to this.
+            let widget = widget_mut
+                .widget
+                .as_mut_dyn_any()
+                .downcast_mut::<Box<dyn Widget>>()
+                .unwrap();
+            let widget_mut = WidgetMut {
+                ctx: widget_mut.ctx.reborrow_mut(),
+                widget,
+                is_reborrow: true,
+            };
+
+            res = Some(f(widget_mut));
+        });
+
+        root_state.merge_up(self.widget_arena.get_state_mut(self.root.id()).item);
+        self.post_event_processing(&mut root_state);
 
+        res.unwrap()
+    }
+
+    /// Get a [`WidgetMut`] to a specific widget.
+    ///
+    /// Because of how `WidgetMut` works, it can only be passed to a user-provided callback.
```

Comment:
If mutations are required to several widgets, is there an inefficiency here? (Question, not looking for a change, I think.)

Reply:
If you mutate several widgets with the same call to `edit_widget`, no. If you call `edit_widget` hundreds of times per frame, yes, a bit.

It's an escape hatch first and foremost, it's not supposed to substitute for the pass system.

Reply:
So when I'm getting it right, later only `edit_root_widget` and `mutate_later` should be used?

Is this method even necessary (outside of tests)? I only see usage of it in the tests, so should this be `pub(crate)` instead?

Reply:
@waywardmonkeys requested it for his use-cases.

It's not necessary for Xilem, but it's easy to implement and I can easily imagine use-cases for it for consumers of Masonry.